commec/setup.py

```diff
@@ -0,0 +1,692 @@
+\033[48;5;17m█ █████▄ █████▄ █ ▀███▄            SETUP                \033[38;5;202m      ▐█████▄██▀    ▀▄    \033[0m   
+\033[48;5;17m█ █    █ █    █ █ ▄   █              UTILITY            \033[38;5;202m      ▐████████       ▌  \033[0m
+\033[48;5;17m█ █████▀ █████▀ █ ▀█▄█▀                                 \033[38;5;202m      ████████▀         \033[0m
+                                                        \033[38;5;202m   ▄▄██████▀▀             \033[0m
+                                                        \033[38;5;202m ▀▀                       \033[0m"""
+        )
+        print(
+            "                 The Common Mechanism!",
+            "\n\nInternational Biosecurity and Biosafety Initiative for Science",
+            "\nCopyright © 2021-2024 ",
+            "\n\nThis script will help download the mandatory databases ",
+            "\nrequired for using Commec Screen, and requires a stable",
+            "\ninternet connection, wget, and update_blastdb.pl.",
+        )
+        self.print_help_info()
+        print()
+        self.check_requirements()
+        self.setup_overall_directory()
+        self.do_setup()
+
+    def print_help_info(self, additional_help=[""]):
+        """
+        Prints helpful instructions according to the current step,
+        as well as general instructions.
+        """
+        add_help_str = "" if len(additional_help) == 0 else "".join(additional_help)
+        print(
+            "\033[38;5;202m"
+            " Instructions: "
+            '\n -> You can exit this setup at any time with "exit"'
+            '\n -> You can return to a previous step with "back"'
+            '\n -> You can get additional help with "help"' + add_help_str + "\033[0m"
+        )
+
+    def check_requirements(self):
+        """Checks for wget, and update_blastdb.pl, which should both be present
+        given the conda environment to install and use Commec."""
+        print("\033[38;5;242mChecking for wget, and update_blastdb")
+        dont_proceed = False
+        result = subprocess.run(["which", "wget"], check=False)
+        if result.returncode != 0:
+            print(
+                "\033[38;5;202mDependency wget could not be found. "
+                "Check wget is installed in this environment."
+            )
+            dont_proceed = True
+        result = subprocess.run(["which", "update_blastdb.pl"], check=False)
+        if result.returncode != 0:
+            print(
+                "\033[38;5;202mDependency update_blastdb.pl could not be found. "
+                "Blast is likely not installed in this python environment."
+            )
+            dont_proceed = True
+        if dont_proceed:
+            print("\033[38;5;202mError: Required dependency missing!")
+            self.stop()
+        print("\033[0m")
+
+    def check_directory_is_writable(self, input_directory: str) -> bool:
+        """Checks a directory is viable by creating it and destroying it."""
```

Comment:
The destruction is exactly to capture the case where the user exits, but we don't want to create the directory yes.

I see the issue, empty files do indeed behave poorly, and disappear if they already existed, which is likely very rarely meaningingful, but not ideal. Perhaps there are better ways to test the validity of a directory without creating it. I'll investigate.

Reply:
Did some chatting with an LLM and I *think* this will work though it does feel kind of horribly overengineered:

```
   try:
        path = os.path.normpath(os.path.abspath(path))
        if os.path.exists(path):
            return os.path.isdir(path)
        
        # We will later use os.makedirs (recursive directory creation) but want to avoid os.rmdirs due
        # to edge cases (pre-existing parent directory was empty), so look for the first directory in the
        # path that os.makedirs would create
        parent_path = path
        while not os.path.exists(parent_path):
            parent_path = os.path.dirname(parent_path)
        
        # Check for errors creating a temp directory in the existing parent directory, then remove it
        temp_dir = os.path.join(parent_path, "tmp_commec_dir")
        os.mkdir(temp_dir)
        os.rmdir(temp_dir)
        return True
```